### PR TITLE
Enable and constrain optimized array operator support for InstantRender

### LINKED ISSUE #2451: Reject aliasing in Array operators
It is desirable, in the case of nested Array operators, to not all-out leak. The contents of the array that are operated upon are leaked on creation. This leaves objects in the environment to deal with. The proposal here is to make all objects transitively reachable via reads to the bindings of optimized array ops "final" so that they are never used again. This solution is not general, but it is easier than the solution proposed in #2449.

The cases to reject include the ones caused by the `this` binding. 

```js
function f(c) {
    var arr = Array.from(c);
    
    let a = {foo:1};
	function op(o) {
		return a;
	}
    
    let mapped = arr.map(op);
    mapped[0].foo = 2;
    let x = a.foo;
    return x;
}

global.__optimize && __optimize(f);

inspect = () => {
  return f([{foo:42}]);
};
```

### PR DESCRIPTION
**Pull request by sb98052**
This change ensures that the code generated via optimized array loop operators is correct in the InstantRender setting. It enables the optimization of such functions when the `--instantRender` option is set. Materialization is prevented by ensuring that post-optimization, objects that are reachable from the function are not mutated. Recoverable errors are issued. We also degrade all InstantRender bailouts to recoverable errors, to facilitate debugging. We add a constraint that optimized functions may not be reused.

Resolves #2451 #2448

### CODE REVIEW

File: src/utils/leak.js [hunk 1]
```diff
@@ -630,19 +630,35 @@ export class MaterializeImplementation {
...
+          o.makeFinal();
+        }
+
+        let error = new CompilerDiagnostic(
+          "Object reached by optimized function may not be subsequently mutated. To enforce this condition, run with --diagnosticAsError PP0044",
+          o.expressionLocation,
+          diagnosticCode,
+          "Warning"
+        );
+        realm.handleError(error);
```
Comment (hermanventer) [inline src/utils/leak.js]: Never ignore the result of handleError and then carry on. You may only ignore it if you throw a FatalError immediately afterwards.

File: src/utils/leak.js [hunk 2]
```diff
@@ -630,19 +630,35 @@ export class MaterializeImplementation {
...
+    let handleMaterialization: ObjectValue => void;
+    if (realm.instantRender.enabled) {
+      // We prevent mutations to objects so that non-final
+      // values cannot occur, and hence materialization is avoided.
+      // The values of properties needed are issued via object literals.
+      // By default, we issue a warning, and enforce the guard only if the
+      // user upgrades the warning to an error.
+      handleMaterialization = o => {
+        let diagnosticCode = "PP0044";
+        if (realm.userChangedDiagnosticToError(diagnosticCode)) {
```
[the remark this answers is not in the document]
Comment (sb98052) [reply src/utils/leak.js]: Ok, I'll find another way to do this.

File: src/utils/leak.js (hunk 1 above)
Comment (sb98052) [reply src/utils/leak.js]: This I've fixed.

**Comment by NTillmann** [comment]
Looks fine to me.

**Comment by sb98052** [comment]
> Looks fine to me; except that I expect to run into issues when the everything including the module table is getting marked as final. But we can deal with it when we get there.

@NTillmann Note that accessing such objects lead to recoverable errors for Instant Render.

File: src/methods/properties.js [hunk 3]
```diff
@@ -231,16 +231,23 @@ function ensureIsNotFinal(realm: Realm, O: ObjectValue, P: void | PropertyKeyVal
...
-  throw new FatalError();
+  // We can't continue because this object is already in its final state
+  if (realm.instantRender.enabled) {
+    realm.instantRenderBailout(
+      "Object mutations that require materialization are currently not supported by InstantRender",
+      realm.currentLocation
+    );
+  } else {
+    let error = new CompilerDiagnostic(
+      "Mutating a final object, or an object with unknown properties, after some of those " +
```
Comment (hermanventer) [inline src/methods/properties.js]: Not your doing, but I hate it when terms related to internal implementation details of Prepack leak out via error messages. Moreover, the text of this message is very obscure and may mislead a programmer who has to make sense of it. 

Even worse than that, the wiki page for this error gives an outdated example. Prepack already snapshots and does the right thing for the example. I'm none too sure if there are other cases. If there are none, the whole concept of final objects should be deleted. Otherwise, the wiki page should be updated with examples that cover ALL of the other cases.

Is there an issue for this?

File: src/realm.js [hunk 4]
```diff
@@ -1655,6 +1656,14 @@ export class Realm {
     return previousValue;
   }
 
+  /* Since it makes strong assumptions, Instant Render is likely to have a large
+  number of unsupported scenarios. We group all associated compiler diagnostics here. */
+  instantRenderBailout(message: string, loc: ?BabelNodeSourceLocation) {
+    if (loc === undefined) loc = this.currentLocation;
+    let error = new CompilerDiagnostic(message, loc, "PP0039", "RecoverableError");
```
Comment (hermanventer) [inline src/realm.js]: As I've said before, we should not use the same error code for different error messages. At best we can parameterize a single message. Given that, this abstraction is a trap and should be removed.

File: src/methods/properties.js [hunk 5]
```diff
@@ -231,16 +231,23 @@ function ensureIsNotFinal(realm: Realm, O: ObjectValue, P: void | PropertyKeyVal
...
-    realm.currentLocation,
-    "PP0026",
-    "FatalError"
-  );
-  realm.handleError(error);
-  throw new FatalError();
+  // We can't continue because this object is already in its final state
+  if (realm.instantRender.enabled) {
+    realm.instantRenderBailout(
+      "Object mutations that require materialization are currently not supported by InstantRender",
```
Comment (hermanventer) [inline src/methods/properties.js]: Materialization is an internal concept and should not leak into error messages. The message is also very obscure. If you can't explain why this is not allowed in clear terms along with a wiki page that has a clarifying example, you are not quite done thinking about this problem.

File: src/serializer/ResidualHeapSerializer.js [hunk 6]
```diff
@@ -1348,14 +1348,10 @@ export class ResidualHeapSerializer {
           to the __empty built-in */
           if (instantRenderMode) {
             if (this.emitter.getReasonToWaitForDependencies(elemVal)) {
-              let error = new CompilerDiagnostic(
+              this.realm.instantRenderBailout(
                 "InstantRender does not yet support cyclical arrays or objects",
```
Comment (hermanventer) [inline src/serializer/ResidualHeapSerializer.js]: This error is totally unrelated to the previous call to instantRenderBailout and a perfect example of the trap that I want to remove.

File: src/values/ArrayValue.js [hunk 7]
```diff
@@ -43,6 +43,11 @@ function evaluatePossibleNestedOptimizedFunctionsAndStoreEffects(
       thisValue = func.$BoundThis;
     }
     invariant(funcToModel instanceof ECMAScriptSourceFunctionValue);
+
+    if (realm.instantRender.enabled && realm.collectedNestedOptimizedFunctionEffects.has(funcToModel)) {
+      realm.instantRenderBailout("Array operators may only be optimized once", funcToModel.expressionLocation);
```
Comment (hermanventer) [inline src/values/ArrayValue.js]: What does this mean? Give an example in a wiki page.

File: src/values/ArrayValue.js [hunk 8]
```diff
@@ -59,6 +64,12 @@ function evaluatePossibleNestedOptimizedFunctionsAndStoreEffects(
       // the default behaviour and leaked the nested functions so any bindings
       // within the function properly leak and materialize.
       if (e instanceof NestedOptimizedFunctionSideEffect) {
+        if (realm.instantRender.enabled) {
+          realm.instantRenderBailout(
+            "InstantRender does not support impure array operators",
```
Comment (hermanventer) [inline src/values/ArrayValue.js]: What is an impure array operator? Give a motivating example in a wiki page.

File: src/serializer/ResidualHeapSerializer.js [hunk 9]
```diff
@@ -1855,14 +1851,10 @@ export class ResidualHeapSerializer {
           let delayReason;
           if (instantRenderMode) {
             if (this.emitter.getReasonToWaitForDependencies(propValue)) {
-              let error = new CompilerDiagnostic(
-                "InstantRender does not yet support cyclical arays or objects",
-                val.expressionLocation,
-                "PP0039",
-                "FatalError"
+              this.realm.instantRenderBailout(
+                "InstantRender does not yet support cyclical arrays or objects",
```
Comment (hermanventer) [inline src/serializer/ResidualHeapSerializer.js]: Why is this no longer a fatal error? Is it likely that the code generated in this case will do the right thing? If so, I want to see a regression test that does the right thing in this case.
